pkg/instrumentation/auto/annotation.go

```diff
@@ -37,25 +37,27 @@ type AnnotationMutators struct {
 	daemonSetMutators   map[string]instrumentation.AnnotationMutator
 	statefulSetMutators map[string]instrumentation.AnnotationMutator
 	defaultMutator      instrumentation.AnnotationMutator
+	injectAnnotations   map[string]struct{}
 }
 
 // RestartNamespace sets the restartedAtAnnotation for each of the namespace's supported resources and patches them.
 func (m *AnnotationMutators) RestartNamespace(ctx context.Context, namespace *corev1.Namespace) {
 	restartAndPatchFunc := m.patchFunc(ctx, setRestartAnnotation)
-	m.rangeObjectList(ctx, &appsv1.DeploymentList{}, client.InNamespace(namespace.Name), restartAndPatchFunc)
-	m.rangeObjectList(ctx, &appsv1.DaemonSetList{}, client.InNamespace(namespace.Name), restartAndPatchFunc)
-	m.rangeObjectList(ctx, &appsv1.StatefulSetList{}, client.InNamespace(namespace.Name), restartAndPatchFunc)
+	checkDifferentInjectAnnotationsFunc := m.checkDifferentInjectAnnotationsFunc(namespace)
+	m.rangeObjectList(ctx, &appsv1.DeploymentList{}, client.InNamespace(namespace.Name), chainCallbacks(checkDifferentInjectAnnotationsFunc, restartAndPatchFunc))
+	m.rangeObjectList(ctx, &appsv1.DaemonSetList{}, client.InNamespace(namespace.Name), chainCallbacks(checkDifferentInjectAnnotationsFunc, restartAndPatchFunc))
+	m.rangeObjectList(ctx, &appsv1.StatefulSetList{}, client.InNamespace(namespace.Name), chainCallbacks(checkDifferentInjectAnnotationsFunc, restartAndPatchFunc))
 }
 
 // MutateAndPatchAll runs the mutators for each of the supported resources and patches them.
 func (m *AnnotationMutators) MutateAndPatchAll(ctx context.Context) {
 	mutateAndPatchFunc := m.patchFunc(ctx, m.MutateObject)
-	m.rangeObjectList(ctx, &corev1.NamespaceList{}, &client.ListOptions{},
-		chainCallbacks(mutateAndPatchFunc, m.restartNamespaceFunc(ctx)),
-	)
 	m.rangeObjectList(ctx, &appsv1.DeploymentList{}, &client.ListOptions{}, mutateAndPatchFunc)
 	m.rangeObjectList(ctx, &appsv1.DaemonSetList{}, &client.ListOptions{}, mutateAndPatchFunc)
 	m.rangeObjectList(ctx, &appsv1.StatefulSetList{}, &client.ListOptions{}, mutateAndPatchFunc)
+	m.rangeObjectList(ctx, &corev1.NamespaceList{}, &client.ListOptions{},
```

Comment:
Flipped the order to mutate the resources first (triggering a restart) and then mutate the namespace such that this time, a 2nd restart can be avoided.